internal/daemon/api_logs.go

```diff
@@ -0,0 +1,275 @@
+		for i, info := range infos {
+			services[i] = info.Name
+		}
+	}
+
+	// Get log iterators by service name (and close them when we're done).
+	itsByName, err := r.svcMgr.ServiceLogs(services, numLogs)
+	if err != nil {
+		response := statusInternalError("cannot fetch log iterators: %v", err)
+		response.ServeHTTP(w, req)
+		return
+	}
+	defer func() {
+		for _, it := range itsByName {
+			_ = it.Close()
+		}
+	}()
+
+	// Output format is JSON Lines, which doesn't have an official mime type,
+	// but "application/x-ndjson" is what most people seem to use:
+	// https://github.com/wardi/jsonlines/issues/9
+	w.Header().Set("Content-Type", "application/x-ndjson")
+	encoder := json.NewEncoder(w)
+	encoder.SetEscapeHTML(false)
+
+	// First write buffered logs
+	if numLogs != 0 {
+		if numLogs < 0 {
+			// Client asked for all logs; write everything from the iterators
+			// (which are TailIterators).
+			err = writeLogs(itsByName, func(entry servicelog.Entry) error {
+				return encoder.Encode(jsonLog(entry))
+			})
+		} else {
+			// Client asked for n logs, collect from iterators (which are
+			// HeadIterator(n)) to a FIFO queue of size n and drop oldest. This
+			// avoids reading numServices*numLogs into memory and sorting.
+			fifo := make(chan servicelog.Entry, numLogs)
+			err = writeLogs(itsByName, func(entry servicelog.Entry) error {
+				// If FIFO channel is full, discard oldest log entry before
+				// writing new one so it doesn't block.
+				if len(fifo) == cap(fifo) {
+					<-fifo
+				}
+				fifo <- entry
+				return nil
+			})
+			// Then write the logs from the FIFO to the response.
+			for len(fifo) > 0 && err == nil {
+				err = encoder.Encode(jsonLog(<-fifo))
+			}
+		}
+		if err != nil {
+			log.Printf("error writing logs: %v", err)
+			return
+		}
+	}
+
+	// If requested, follow and output logs real time until request closed.
+	if follow {
+		flushWriter(w) // first flush any logs written above
+
+		// followLogs sends logs to output channel as they arrive. Do this
+		// until request is cancelled.
+		output := make(chan servicelog.Entry)
+		go followLogs(itsByName, req.Context().Done(), output)
+		for entry := range output {
+			err = encoder.Encode(jsonLog(entry))
+			if err != nil {
+				log.Printf("error writing logs: %v", err)
+				return
+			}
+			flushWriter(w)
+		}
+	}
+}
+
+// writeLogs reads logs from the given iterators, merging the log streams and
+// calling writeLog for each log.
+func writeLogs(itsByName map[string]servicelog.Iterator, writeLog writeLogFunc) error {
+	// Stop when we get a log recorded after this time. This prevents
+	// continually trying to catch up if the client is slow at reading them.
+	stop := time.Now().UTC()
+
+	// Service names, ordered alphabetically for consistent output.
+	names := make([]string, 0, len(itsByName))
+	for name := range itsByName {
+		names = append(names, name)
+	}
+	sort.Strings(names)
+
+	parsersByName := make(map[string]*servicelog.Parser, len(itsByName))
```

Comment:
parser setup should happen outside of writeLogs so that writeLogs and followLogs can share the same parser so to not skip data already buffered.

Reply:
Done.